### PR TITLE
Timeout on HTML Validator

### LINKED ISSUE #145: Timeout on validator initalization
We should set a good chunk of time before we decide that for some reason html validator initialization failed and then just print an error to the user saying validation is disabled because of a problem and move on. Otherwise the app will hang and never start because right now the validation initialization is blocking; if it never completes the app never finishes starting

Maybe something like: Wait 30 seconds or so, provide the warning, and print standard out from the `exec` (so that they can file an issue with that or look into things further?)

Related to #143 and #135

### PR DESCRIPTION
**Pull request by JoshL18**
(closes #145)

Created a 10 second timeout on the HTML Validator to ensure whether the validation completed. Initialization continues even in the case that the validator has been disabled. Errors during validation initialization are now printed to console in red for better visibility.

### CODE REVIEW

File: lib/htmlValidator.js [hunk 1]
```diff
@@ -22,8 +24,9 @@ module.exports = function(app, callback) {
       htmlArrLength,
       htmlLine,
       formattedHTML,
+      validatorTimeout,
       errorPage = path.normalize(__dirname + '/../defaultErrorPages/views/validatorErrors'),
-      validatorDisabled;
+      validatorDisabled = true;
```
Comment (Autre31415) [inline lib/htmlValidator.js]: This shouldn't need to be hardcoded. It's taken care of in the following if statement.

File: lib/htmlValidator.js [hunk 2]
```diff
@@ -48,29 +51,38 @@ module.exports = function(app, callback) {
 
         validatorProcess.stdout.on('data', (data) => {
           if (`${data}`.includes('Initialization complete')) {
+            validatorDisabled = false;
```
Comment (Autre31415) [inline lib/htmlValidator.js]: This line becomes redundant when you remove the hardcoding of this variable.

File: lib/htmlValidator.js [hunk 3]
```diff
@@ -48,29 +51,38 @@ module.exports = function(app, callback) {
...
             reject();
           }
         });
+
+        validatorTimeout = setTimeout(function() {
+          if (validatorDisabled) {
+            console.error(('HTML Validator has timed out.').red);
+            reject();
+          }
+        }, 10000);
```
Comment (Autre31415) [inline lib/htmlValidator.js]: Are we sure that 10 seconds is enough for some slower machines or laggy vms?

File: lib/htmlValidator.js [hunk 4]
```diff
@@ -160,7 +172,11 @@ module.exports = function(app, callback) {
     next();
   });
 
-  Promise.all([promise]).then(() => {
+  promise.then(function (reason) {
```
Comment (Autre31415) [inline lib/htmlValidator.js]: Remove the space between the function and arg, and rename arg to result

File: lib/htmlValidator.js [hunk 5]
```diff
@@ -160,7 +172,11 @@ module.exports = function(app, callback) {
...
+    callback();
+  }).catch(function (reason) {
```
Comment (Autre31415) [inline lib/htmlValidator.js]: Same here

File: lib/htmlValidator.js [hunk 6]
```diff
@@ -48,29 +51,38 @@ module.exports = function(app, callback) {
...
         });
 
         validatorProcess.stderr.on('data', (data) => {
           if (`${data}`.includes('No Java runtime present, requesting install.')) {
-            console.error('error: No Java runtime detected. Check to make sure Java is installed and in your path.');
+            console.error(('There was an error in authentication. Please make sure to have Java installed.').red);
+            validatorDisabled = true;
             reject();
           }
-          else {
```
Comment (alallier) [inline lib/htmlValidator.js]: Don't mean to step on your toes @Autre31415, I have just been helping @JoshL18 in person with this. You have to be careful with the removal of this else, (is there a case when we will get something in standard error and the validator can't start? If so we will want this else.

I may have told you to remove this @JoshL18, if I did sorry about that.

Comment (Autre31415) [inline lib/htmlValidator.js]: As far as I'm aware anything we're getting in standard error should cause the promise to be rejected. We're only resolving the promise when we know for sure that the validator is working as intended.

Would be a good idea to test this though. A console dump of all data under stderr should clarify.

Comment (Autre31415) [inline lib/htmlValidator.js]: As far as I'm aware anything we're getting in standard error should cause the promise to be rejected. We're only resolving the promise when we know for sure that the validator is working as intended.

Would be a good idea to test this though. A console dump of all data under stderr should clarify.

Comment (alallier) [inline lib/htmlValidator.js]: Could there be a race condition, if we get some standard error, but everything still initiates, resulting in a reject and then a resolve?

Comment (Autre31415) [inline lib/htmlValidator.js]: Once a promise is handled it's done. It won't be handled more than once.

File: lib/htmlValidator.js (hunk 4 above)
Comment (alallier) [inline lib/htmlValidator.js]: Btw @Autre31415 `reason` seems to be the standard according to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch) not sure if that matters as far as readability/styling is concerned?

File: lib/htmlValidator.js (hunk 6 above)
Comment (Autre31415) [inline lib/htmlValidator.js]: Alex and I have discussed this and we're in agreement this else should be left out.

File: lib/htmlValidator.js [hunk 7]
```diff
@@ -46,31 +49,43 @@ module.exports = function(app, callback) {
           'java', ['-Xss1024k', '-cp', vnu, 'nu.validator.servlet.Main', params.htmlValidator.port ? params.htmlValidator.port : '8888']
         );
 
+        console.log(('Validator intializing...').yellow);
```
Comment (JoshL18) [inline lib/htmlValidator.js]: @kethinov Is the wording here okay?

File: lib/htmlValidator.js [hunk 8]
```diff
@@ -46,31 +49,43 @@ module.exports = function(app, callback) {
...
+
         validatorProcess.stdout.on('data', (data) => {
           if (`${data}`.includes('Initialization complete')) {
+            clearTimeout(validatorTimeout);
+            console.log(('Validator initialization completed successfully.').green);
```
Comment (JoshL18) [inline lib/htmlValidator.js]: @Autre31415 I added a line to tell the user when initialization was successful, should we keep this or just get rid of it?

File: lib/htmlValidator.js (hunk 7 above)
Comment (kethinov) [inline lib/htmlValidator.js]: Go with "Starting validator..."

Comment (kethinov) [inline lib/htmlValidator.js]: Actually make that "Starting HTML validator..."

File: lib/htmlValidator.js (hunk 8 above)
Comment (kethinov) [inline lib/htmlValidator.js]: And "HTML validator listening on port [port]"

use bold for both messages

Comment (kethinov) [inline lib/htmlValidator.js]: and yeah we only need one message indicating that the validator was started, so pick the line of code that makes the most sense for it to log it and use that message

File: lib/htmlValidator.js [hunk 9]
```diff
@@ -46,31 +49,43 @@ module.exports = function(app, callback) {
...
             resolve();
           }
         });
 
         validatorProcess.stderr.on('data', (data) => {
           if (`${data}`.includes('No Java runtime present, requesting install.')) {
-            console.error('error: No Java runtime detected. Check to make sure Java is installed and in your path.');
+            console.error(('There was an error in authentication. Please make sure to have Java installed.').red);
+            validatorDisabled = true;
```
Comment (Autre31415) [inline lib/htmlValidator.js]: Instead of setting this to true on each error, you could set it to true in the promise rejection catch.

**Comment by kethinov** [comment]
Emojis! 🎉🎊🎉🎊🎉🎊

**Pull request merged**